plugins/incremental-ingestion-backend/src/database/IncrementalIngestionDatabaseManager.ts

```diff
@@ -0,0 +1,592 @@
+
+      return {
+        deletions: {
+          markEntitiesDeleted,
+          marksDeleted,
+          ingestionsDeleted,
+        },
+      };
+    });
+  }
+
+  /**
+   * Automatically cleans up duplicate ingestion records if they were accidentally created.
+   * Any ingestion record where the `rest_completed_at` is null (meaning it is active) AND
+   * the ingestionId is incorrect is a duplicate ingestion record.
+   * @param ingestionId - string
+   * @param provider - string
+   */
+  async clearDuplicateIngestions(ingestionId: string, provider: string) {
+    await this.client.transaction(async tx => {
+      const invalid = await tx<IngestionRecord>('ingestion.ingestions')
+        .where('provider_name', provider)
+        .andWhere('rest_completed_at', null)
+        .andWhereNot('id', ingestionId);
+
+      if (invalid.length > 0) {
+        await tx('ingestion.ingestions').delete().whereIn('id', invalid);
+        await tx('ingestion.ingestion_mark_entities')
+          .delete()
+          .whereIn(
+            'ingestion_mark_id',
+            tx('ingestion.ingestion_marks')
+              .select('id')
+              .whereIn('ingestion_id', invalid),
+          );
+        await tx('ingestion.ingestion_marks')
+          .delete()
+          .whereIn('ingestion_id', invalid);
+      }
+    });
+  }
+
+  /**
+   * This method fully purges and resets all ingestion records for the named provider, and
+   * leaves it in a paused state.
+   * @param provider - string
+   * @returns Counts of all deleted ingestion records
+   */
+  async purgeAndResetProvider(provider: string) {
+    return await this.client.transaction(async tx => {
+      const ingestionIDs: { id: string }[] = await tx('ingestion.ingestions')
+        .select('id')
+        .where('provider_name', provider);
+
+      const markIDs: { id: string }[] =
+        ingestionIDs.length > 0
+          ? await tx('ingestion.ingestion_marks')
+              .select('id')
+              .whereIn(
+                'ingestion_id',
+                ingestionIDs.map(entry => entry.id),
+              )
+          : [];
+
+      const markEntityIDs: { id: string }[] =
+        markIDs.length > 0
+          ? await tx('ingestion.ingestion_mark_entities')
+              .select('id')
+              .whereIn(
+                'ingestion_mark_id',
+                markIDs.map(entry => entry.id),
+              )
+          : [];
+
+      const markEntitiesDeleted = await this.deleteMarkEntities(
+        tx,
+        markEntityIDs,
+      );
+
+      const marksDeleted =
+        markIDs.length > 0
+          ? await tx('ingestion.ingestion_marks')
+              .delete()
+              .whereIn(
+                'ingestion_id',
+                ingestionIDs.map(entry => entry.id),
+              )
+          : 0;
+
+      const ingestionsDeleted = await tx('ingestion.ingestions')
+        .delete()
+        .where('provider_name', provider);
```

Comment:
yeah a lot of the cleanup would go away if you had on cascade delete foreign relations as suggested in the migration

Reply:
I couldn't find a way to do this without losing the information about how many of each type of record was deleted. I'm still trying to decide if I actually care about that data, though.